lnwire/dyn_reject.go

```diff
@@ -0,0 +1,220 @@
+package lnwire
+
+import (
+	"bytes"
+	"io"
+
+	"github.com/btcsuite/btcd/btcec/v2"
+	"github.com/btcsuite/btcd/btcutil"
+	"github.com/lightningnetwork/lnd/tlv"
+)
+
+// DynReject is a message that is sent during a dynamic commitments negotiation
+// process. It is sent by both parties to propose new channel parameters.
+type DynReject struct {
+	// ChanID identifies the channel whose parameters we are trying to
+	// re-negotiate.
+	ChanID ChannelID
+
+	// DustLimit, if not nil, proposes a change to the dust_limit_satoshis
+	// for the sender's commitment transaction.
+	DustLimit *btcutil.Amount
+
+	// MaxValueInFlight, if not nil, proposes a change to the
+	// max_htlc_value_in_flight_msat limit of the sender.
+	MaxValueInFlight *MilliSatoshi
+
+	// ChannelReserve, if not nil, proposes a change to the
+	// channel_reserve_satoshis requirement of the recipient.
+	ChannelReserve *btcutil.Amount
+
+	// CsvDelay, if not nil, proposes a change to the to_self_delay
+	// requirement of the recipient.
+	CsvDelay *uint16
+
+	// MaxAcceptedHTLCs, if not nil, proposes a change to the
+	// max_accepted_htlcs limit of the sender.
+	MaxAcceptedHTLCs *uint16
+
+	// FundingKey, if not nil, proposes a change to the funding_pubkey
+	// parameter of the sender.
+	FundingKey *btcec.PublicKey
```

Comment:
I get that this is here to be identical to `DynPropose`, but I don't think a counterparty should suggest what funding_pubkey we should use

Reply:
It's not supposed to be a suggestion, but a problem with the specific change. I do agree though that they shouldn't be able to reject the funding key. If they don't accept the one presented then they can reject the whole negotiation.

Reply:
https://github.com/ProofOfKeags/bolts/blob/extension/dynamic-commitments/ext-dynamic-commitments.md#dyn_reject

Reply:
It seems like the responder has to set this field to something if they reject it, which is a little confusing since that means they have to populate this pubkey

Reply:
Again, the record is just supposed to be copied from the propose to the reject unchanged if it is unsatisfactory. It is not supposed to be a countersuggestion.

Think about it this way: "these are the specific fields I have a problem with" NOT "please send these fields instead".

 I toyed with the idea instead to just have a bit vector where the bit positions that were set line up with the tlv type numbers it rejects, therefore removing that field altogether, but I figured this would be easier to implement and explain, since it would be one less mechanism to design. It appears I was wrong.

Now that I think about it more, it does leave some UB with regards to what happens if the field is rejected but the value in the field sent back differs from the one that is sent out.

Do you think I should fix this by reworking the structure or just do a better job of explaining it in the spec? What do you think about the idea to just have a bit vector where the responder only needs to express a binary choice between accept and reject of the parameter?

Reply:
IMO the structure should be reworked. To me it seems that there are two ways to go:
- the bitvector approach (likely a TLV so it doesn't have a fixed size so we can expand on it in the future)
- a TLV for each field (excluding the `funding_pubkey`) similar to what's done here. However, in the case of numerical fields "too high" or "too low" can be returned in the TLV to point the proposer to more appropriate values. For non-numerical fields like the `channel_type`, I'm not sure what can be done besides sending back something that denotes "bad value".  This option seems a little more ambitious